src/util/README.md

```diff
@@ -101,19 +101,90 @@ of a while loop is a statement.
 
 \subsection symbolt_section symbolt, symbol_tablet, and namespacet
 
-To be documented.
-
-\subsubsection symbol_lifetimes_section Symbol lifetimes, symbol modes, name, base-name, pretty-name; semantic of lifetimes for symex?
-
-To be documented.
-
-\subsubsection storing_symbols_section Storing symbols and hiding symbols (namespacet)
-
-To be documented.
-
-\subsubsection ns_follow_section ns.follow
-
-To be documented.
+A symbol table is a mapping from symbol names to \ref symbolt objects, which
+store a symbol's name, attributes, type and perhaps value. They are used to
+describe local and global variables, type definitions and function prototypes
+and definitions.
+
+All symbols store a type (an instance of \typet). For function or method symbols
+these are \ref code_typet instances.
+
+Global variable symbols may have a value (an \ref exprt), in which case it is
+used to initialise the global. Method or function symbols may also have a value,
+in which case it is a \ref codet and gives the function definition. A method or
+function symbol without a value is a prototype (for example, it might be an
+`extern` declaration in C). Local variables' symbol values are always ignored;
+any initialiser must be explicitly assigned after they are instantiated by a
+declaration (\ref code_declt).
+
+Symbol expressions (\ref symbol_exprt) and types (\ref symbol_typet) refer to
+symbols stored in a symbol table. Symbol expressions can be thought of as
+referring to the table for more detail about a symbol (for example, is it a
+local or a global variable, or perhaps a function?), and have a type which must
+match the type given in the symbol table. Symbol types can be thought of as
+shorthands or aliases for a type given in more detail in the symbol table, for
+example permitting a shorthand for a large structure type, as well as permitting
+the construction of expressions referring to that type before its full
+definition is known.
+
+Note the implementation of \ref symbol_tablet is split into a base interface
+(\ref symbol_table_baset) and an implementation (\ref symbol_tablet). There is
+one alternate implementation (\ref journalling_symbol_tablet) which additionally
+maintains a log or journal of symbol creation, modification and deletions.
+
+Namespaces (\ref namespacet) provide a read-only view on one or more symbol
+tables, and provide helper functions that aid accessing the table. A namespace
+may layer exactly two symbol tables, in which case any lookup operation checks
```

Comment:
A namespace may have a single symbol table, or also many more (`multi_namespacet`).

Reply:
so there is... in that case why on earth can namespacet do either 1 or 2 layers? Just backward compatibility?

Reply:
I'm not sure "1 or 2 layers" is a wording I'd use, but we've got:
1. Most cases just have a `namespacet` backed by a single symbol table.
2. goto-symex (and possibly others) uses a second symbol table to store symbols introduced during symbolic execution, which are otherwise not of interest and thus shouldn't result in modification of the original symbol table.
3. I am not aware of any use of `multi_namespacet`.